### PR TITLE
Advanced Building Blocks - Enumerable Methods[SOLO]

### PR DESCRIPTION
**Pull request by tirthajyoti-ghosh**
- Built-in enum methods such as: `#each`, `#each_with_index`, `#select`, `#all?`, `#any?`, `#none?`, `#count`, `#map`, `#inject`  re-written in custom algorithm.
- Made separate file for testing - `enum_test.rb`. `enumerables.rb` file containing actual method codes.
- Method names are like this: `my_` followed built-in method names.
- `#multiply_els` created in `enum_test.rb` file to test `#my_inject` as per project requirements.

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Progress :+1: 
You have made amazing progress with the project. Your code covers most edge cases but we need to fix a few failing ones cases before we can merge.
Check my comments for suggestions.

### Martin Afani

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  def my_all?
+    i = 0
+    while i < size
+      return false if block_given? && !yield(self[i])
+      return false if self[i] == false || self[i].nil?
+
+      i += 1
+    end
+    true
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9) }
array.my_all?(Integer) == array.all?(Integer)
```
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
words = %w[dog door rod blade]
words.my_all?(/d/) == words.all?(/d/)
```
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
array.my_all?(3) == array.all?(3) #true

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  def my_any?
+    i = 0
+    while i < size
+      return true if block_given? && yield(self[i])
+      return true if !self[i] == false && !self[i].nil?
+
+      i += 1
+    end
+    false
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
true_array = [nil, false, true, []] 
false_array = [nil, false, nil, false]
true_array.my_any? == true_array.any? #true
```
- [ ]  when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
words = %w[dog door rod blade]
words.my_any?(Integer) == words.any?(Integer) #true
```
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
words.my_any?(/z/) == words.any?(/z/) #true
```
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
words.my_any?('cat') == words.any?('cat') #true

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      if block_given?
+        return false if yield(self[i])
+      elsif !self[i] == false && !self[i].nil?
+        return false
+      end
+
+      i += 1
+    end
+    true
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9) }
array.my_none?(String) == true #true
```
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
words = %w[dog door rod blade]
words.my_none?(5) == words.none?(5) #true
```
- [ ] Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
 words.my_none?(/z/) == words.none?(/z/) #true
 words.my_none?(/d/) == words.none?(/d/) #true
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                      yield(accumulator, array[i])
+                    elsif param2.nil?
+                      accumulator.send(param1, array[i])
+                    else
+                      accumulator.send(param2, array[i]).send(param2, param1)
+                    end
+      i += 1
+    end
+    accumulator
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) #true
```

Comment (tirthajyoti-ghosh) [reply enumerables.rb]: `array = Array.new(100) { rand(0...9) }`
`array.my_inject(:+) == array.inject(:+) #true`
The code above returns to true in my machine without any changes done after code review.

**Comment by tirthajyoti-ghosh** [comment]
Thanks, @whiz25 for the extensive code review.

Changes are done after code review:
- added Regexp, Class and Pattern support in #my_all?, #my_any?, #my_none?

*Changes requested on #my_inject method is already done and the test-case suggested returns to `true`, so no changes done there

**Review by MauricioRobayo — CHANGES_REQUESTED** [review]
Very nice project!

It looks great and is nicely coming into place, but still need to fix some issues before continuing:

* [ ] `my_inject` is not behaving as `inject`, for example: `[1, 2, 3].my_inject(3, :*) == [1, 2, 3].inject(3, :*)`.
* [ ] `my_any?` should return false when no block or argument is given if the collection only contains false or nil.

Status: __Request changes__.

**Comment by tirthajyoti-ghosh** [comment]
Thanks, @MauricioRobayo for pointing out the above failing test cases.

These are the changes done after the 2nd code review:
- Fixed #my_inject to properly compute 2 params when no block is given.
- Fixed #my_any? to return false when no block or param is given if the collection only contains false or nil.
- Added range support in #my_each.

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello @tirthajyoti-ghosh  ,  You have done a great job with this project. :clap:

**Pull request merged**